### ISSUE TITLE
User should be able to unbond & withdraw funds

### ISSUE THREAD

**Issue by jacogr** [polkadot-js/apps#859]
(Split from #715 with the outstanding items)

Goal:

Allow the user to properly manage their funds in the staking module and allow them to move them between accounts. 

Background:

Currently the user can bond, add additional funds, but cannot withdraw. 

Pre-requisites:

A good understanding of the functions available inside the staking module https://polkadot.js.org/api/METHODS_EXTRINSICS.html#staking

Actions:

(All these apply to the controller)

- Allow the user to unbond, via `unbond` - clicking creates a pop-up that allows the suer to specify the amount he/she wants to unbond. (This should be checked against the available)
- Allow the user to withdraw via `withdrawUnbonded` - this option should _only_ be available when there are funds that have been unbonded already
- Additionally, when the user has funds that have been unbonded, but are not available yet - show the information for these on the overview for the account - fund total and when (block) these will be available. (Information available from the staking ledger)

**Comment by jacogr** [comment]
The tricky bits here are really UX - we don't want to end up with a deluge of buttons. Probably want to re-think things here a bit and have 2 sub-sections, one for the actual actions such as "setting keys" (also see <to be logged>) and one for fund management.

**Comment by Tbaut** [comment]
I've been thinking about it in the past days, studied the apis and want to take a stab at it.

>- Allow the user to unbond, via unbond - clicking creates a pop-up that allows the suer to specify the amount he/she wants to unbond. (This should be checked against the available)

Yes and remove the "Bound additional" when there's nothing more you can add from the stash.


>- Allow the user to withdraw via withdrawUnbonded - this option should only be available when there are funds that have been unbonded already

I was talking with @kianenigma to see if this can be automated on the backend side as this is a pretty bad UX that's not understandable by users. 

> -Additionally, when the user has funds that have been unbonded, but are not available yet - show the information for these on the overview for the account - fund total and when (block) these will be available. (Information available from the staking ledger)

I couldnt'd come with something nicer than:
![image](https://user-images.githubusercontent.com/33178835/56821767-87b4f800-684f-11e9-990f-f52f922018aa.png)

**Comment by Tbaut** [comment]
Circling back on the discussion I had with @Kianenigma. The reason to have a `withdrawUnbonded` is to release the workload on the engine, and move this hurdle to the UI (the hurdle being: check whenever some funds are free to be released and call the extrinsic `withdrawUnbonded`). We will therefore not be able to push back on that.

I suggest to show `available` funds instead of `balance` in any overview (along with`bonded`) and add a `locked` amount of funds and how long it'll be locked for in blocks, see previous post (edited). the drawback is that it makes the funds hard to read (many lines).

To release / unlock the funds:
1. We can add a `Unlock` button or similar in the staking account actions that will only appear when actionable.

=> Drawback: not automated (user will need to click on a new `Unlock` button in the staking area).

2. Another idea is to semi-automate the release of funds whenever it is the right time, for this we can
- show the lock funds as `available` as soon as the time has come.
- as soon as a transfer / bonding action is requested on the stash, let the controller first release the funds before do anything with the stash.

=> Drawback: any lock fund can actually be slashed, hiding this from the user seems not fair.

Both ideas are not ideal. I'll start implementing the 1st that seems to be easier although probably a little less user-friendly. 

Any other idea?

**Comment by jacogr** [comment]
I agree with option 1 here (and yes, understand the "why" behind the node logic).

 I'd rather have the functionality available (even if not perfect) - actually using it (and getting others to use it) is the first hurdle. Once we are actually able to use the damn stuff, it is much easier to plot which part works, what additionally is needed and what just flat out does not work.

**Comment by Kianenigma** [comment]
As mentioned in person I also think it is _safer_ for the time being to implement things in a bit user-unfriendly way but as they are designed to function. UX optimization can be applied later. 

In light of the whole lock/balance thing and how it affects staking: Would be nice if the UI had someplace where you could exactly see the details of your balance. As we know it is way more than just a number. Your total balance can be composed of: free, reserved or vesting. On top of all of these, locks could be applied to your balance. Would be helpful to see them all in one place, and now relating again to this issue: have this `withraw_unbonded()` call next to your `locks` list, if you have any.  

If this is a reasonable feature I can make a separate issue for it?

**Comment by jacogr** [comment]
@Kianenigma The different balance split certainly is more than a reasonable thing to ask for. (Reserved actually does, or used to show up along with balance - however, without side-tracking here, actually had this convo this morning and I'm not 100% that the split in poc-3 didn't break that)

TL;DR - Yes, we should have an issue to properly breakdown the balances (at least for just the staking is is not done at all).
